From f5c6c738115b2be7504f68c7d3220e47895ba632 Mon Sep 17 00:00:00 2001
From: Yann Dirson <yann.dirson@vates.tech>
Date: Wed, 15 Oct 2025 14:01:08 +0200
Subject: [PATCH] Use memory barriers defined by xen headers

This gives us the necessary definitions for arm64.

Those live in xen-barrier.h starting with 4.20, but we won't be able
to use it for some time.  Before this they live in xenctrl.h, but it
pulls too much stuff.  So this is a sync with the 4.20 headers
contents.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>
---
 include/blktap-xenif.h | 24 ++++++++++++++++++++----
 1 file changed, 20 insertions(+), 4 deletions(-)

diff --git a/include/blktap-xenif.h b/include/blktap-xenif.h
index da0deef..15a882c 100644
--- a/include/blktap-xenif.h
+++ b/include/blktap-xenif.h
@@ -14,12 +14,28 @@
 
 /* Get the correct defines for memory barriers - do not fall
  * back to those provided by the kernel */
-#ifdef __i386__
+/* from <xen-barrier.h> in 4.20.0 */
+
+#define xen_barrier() asm volatile ( "" : : : "memory")
+
+#if defined(__i386__)
 #define xen_mb()  asm volatile ( "lock addl $0, -4(%%esp)" ::: "memory" )
-#else
+#define xen_rmb() xen_barrier()
+#define xen_wmb() xen_barrier()
+#elif defined(__x86_64__)
 #define xen_mb()  asm volatile ( "lock addl $0, -32(%%rsp)" ::: "memory" )
+#define xen_rmb() xen_barrier()
+#define xen_wmb() xen_barrier()
+#elif defined(__arm__)
+#define xen_mb()   asm volatile ("dmb" : : : "memory")
+#define xen_rmb()  asm volatile ("dmb" : : : "memory")
+#define xen_wmb()  asm volatile ("dmb" : : : "memory")
+#elif defined(__aarch64__)
+#define xen_mb()   asm volatile ("dmb sy" : : : "memory")
+#define xen_rmb()  asm volatile ("dmb sy" : : : "memory")
+#define xen_wmb()  asm volatile ("dmb sy" : : : "memory")
+#else
+#error "Define barriers"
 #endif
-#define xen_rmb() asm volatile ("" ::: "memory")
-#define xen_wmb() asm volatile ("" ::: "memory")
 
 #endif /* __BLKTAPXENIF_H__ */
-- 
2.47.3

